### ISSUE TITLE
Stochasticity after reshaping operations

### ISSUE THREAD

**Issue by khanhptnk** [pytorch/pytorch#1311]
I would love to see PyTorch maintaining Stochasticity of the output after reshaping. Right now, if you concat or just .view(-1) a stochastic node, you can't backprop by .reinforce(), which is not convenient for structure prediction like machine translation.

**Comment by jekbradbury** [comment]
You can always work around it by reshaping, concatting, or splitting your reward but I agree that this would be nice.

**Comment by khanhptnk** [comment]
Sorry I don't follow your suggestion. You can concat the rewards but you have to concat the decisions as well to match the shape, which loses stochasticity. My most desirable use case to concat the decisions over many time steps and backprop once for efficiency. Correct me if I am wrong.

**Comment by jekbradbury** [comment]
For now you have to keep the decisions in a list, then iterate over it and call `.reinforce` on each one.

**Comment by khanhptnk** [comment]
Right, but there are two problems if your model is an RNN:
1. Because the decisions are computed by shared RNN parameters. You would need to use retain_variables so that you can backward() multiple times through a parameter. And I guess for the last backward(), you have to switch retain_variables off so that those buffers are destroyed. 
2. If all backward() are conducted independently, then each node in the computation graph is visited more than once, which is inefficient. If you can concat and backward() at once, then each node is only visited once. 

I think those are the reasons I feel inconvenient. Right now I am not using reinforce(), but just specifying a loss that incurs the same gradients as doing reinforce(), which allows me to concat and backward() all steps at once.

**Comment by khanhptnk** [comment]
To be clear about why I can't compute rewards and backward() at each time step. My rewards are only given at the end of the sentence.

**Comment by jekbradbury** [comment]
Ah, ok, this is a common situation with an existing (but undocumented) best practice. You want to conduct only one `.backward`, but you have several stochastic nodes. So you can take the stochastic nodes, call `.reinforce` on each of them, then add/concat/otherwise combine them (I use `sum(list_of_stochastic_nodes)` and backpropagate from `var = 0 * sum(list_of_stoch_nodes); var.backward()`. If you also have a non-RL loss, you can add this `0*` term to the loss before calling `.backward`.

**Comment by jekbradbury** [comment]
You can also use `autograd.backward` with two arguments: the list of stochastic nodes and a list of `[None] * len(stochastic_nodes)`

**Comment by khanhptnk** [comment]
ah, great, thank you :)

**Comment by willwhitney** [comment]
The inability to reinforce subsets of stochastic nodes is still a real annoyance in other cases.

The fact that 

```python
sample = a.multinomial()
sample[0].reinforce(1)
```

doesn't work is pretty surprising given that autograd works with indexing in every other situation.

**Comment by jekbradbury** [comment]
Yeah, for now you have to essentially do manual backprop to get the reward tensor you need to submit to the stochastic node.

**Comment by soumith** [comment]
stochastic nodes  no longer exist, so this can be closed.
